File: leetcode/104.maximum-depth-of-binary-tree.py
#!/usr/bin/env python3
'''
leetcode #104 maximum-depth-of-binary-tree 二叉树的最大深度
https://leetcode-cn.com/problems/maximum-depth-of-binary-tree/description/
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findDepth(self, root, n):
        if not root:
            return n
        return max(self.findDepth(root.left, n + 1), self.findDepth(root.right, n + 1))

    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.findDepth(root, 0)

File: leetcode/104.maximum-depth-of-binary-tree.go
package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	if root.Right == nil && root.Left == nil {
		return 1
	}
	if root.Right == nil {
		return 1 + maxDepth(root.Left)
	}
	if root.Left == nil {
		return 1 + maxDepth(root.Right)
	}
	return 1 + max(maxDepth(root.Right), maxDepth(root.Left))
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
